video: ov7675: add Omnivision OV7675 sensor support and basic controls

mjs513 opened the pull request:
The Omnivision OV7675 sensor is currently not supported in Zephyr however it is needed for supporting the Arduino GIGA which currently supports the sensor.  The driver is based on the other sensors primarily the ov7670, gc2145, and ov2640 etc.

The register settings were copied from the[ arducam_dvp library in the arduino repository](https://github.com/ArduCAM/Arducam_dvp).

The camera was tested on the Arduino GIGA display shield, the portenta h7 using a couple of different methods.  Video to giga display or a ST7796 and a single bitmap capture to a Arducam camera app for windows.

See https://github.com/zephyrproject-rtos/zephyr/discussions/93058#discussioncomment-13935281

Not sure what else is needed - really new at doing PRs for zephyr.

mjs513 committed:
add ov7675 driver

mjs513 commented:
@avolmat-st 
> Having looked a bit a the driver ov7670 and ov7675, at least some (I only check few of them) of the registers seems to be same between the two sensors (which we could expect, having numbering quite close).
>Actually I am wondering if having two different drivers for those sensors is necessary or if both sensors could be handled by the very same driver.

To be honest.  That was the way we did it (1 driver) when we put together the [Teensy_Camera library](https://github.com/mjs513/Teensy_Camera).

Unfortunately according to the specs both the OV7675 and the OV7670 both have the same Product ID (0x76).  Yeah I know know kind of strange.  We got around that by passing in a variable to differentiate between the 2 in the begin statement.  So not sure how you would handle that.

Even in the Arducm_DVP library for the Arduino there is a separate class for the OV7675 that uses the OV7670 class as a friend class.   While the default registers are the same there are differences in the resolution tables.

Could you use the CONFIG_VIDEO_OV7675 = y as a differntiator?  Still learning zephyr so just guessing.  But have a feeling might get messy.

Mike

avolmat-st commented:
> @avolmat-st
> 
> > Having looked a bit a the driver ov7670 and ov7675, at least some (I only check few of them) of the registers seems to be same between the two sensors (which we could expect, having numbering quite close).
> > Actually I am wondering if having two different drivers for those sensors is necessary or if both sensors could be handled by the very same driver.
> 
> To be honest. That was the way we did it (1 driver) when we put together the [Teensy_Camera library](https://github.com/mjs513/Teensy_Camera).
> 
> Unfortunately according to the specs both the OV7675 and the OV7670 both have the same Product ID (0x76). Yeah I know know kind of strange. We got around that by passing in a variable to differentiate between the 2 in the begin statement. So not sure how you would handle that.
> 
> Even in the Arducm_DVP library for the Arduino there is a separate class for the OV7675 that uses the OV7670 class as a friend class. While the default registers are the same there are differences in the resolution tables.
> 
> Could you use the CONFIG_VIDEO_OV7675 = y as a differntiator? Still learning zephyr so just guessing. But have a feeling might get messy.
> 
> Guess once this is decided I can make it finally.
> 
> Oh --- is there a reference on what should go into the yaml for the camera?
> 
> Mike

Yeah you can use the compatible as a differentiator for that and within the driver based on the compatible do one thing or the other. There could be several way to do that.

Just grepping rapidly within the Zephyr source code, here's an example of such case:
drivers/adc/adc_ads1x4s0x.c

In this file you can for example see things like:
```
#define ADS1X4S0X_HAS_24_BIT_DEV                                                                   \
        (DT_HAS_COMPAT_STATUS_OKAY(ti_ads124s06) || DT_HAS_COMPAT_STATUS_OKAY(ti_ads124s08))
```
Which is checking for the compatibe ti,ads124s06 or ti,ads124s08

Also doing things such as below to initialize the driver in different way depending on the compatible found:
```
/*
 * ADS124S06: 24 bit, 6 channels
 */
#undef DT_DRV_COMPAT
#define DT_DRV_COMPAT ti_ads124s06
#if DT_HAS_COMPAT_STATUS_OKAY(DT_DRV_COMPAT)
#define ADC_ADS124S06_INST_DEFINE(n) ADC_ADS1X4S0X_INST_DEFINE(n, ti_ads124s06, 6, 24)
DT_INST_FOREACH_STATUS_OKAY(ADC_ADS124S06_INST_DEFINE);
#endif

/*
 * ADS124S08: 24 bit, 12 channels
 */
#undef DT_DRV_COMPAT
#define DT_DRV_COMPAT ti_ads124s08
#if DT_HAS_COMPAT_STATUS_OKAY(DT_DRV_COMPAT)
#define ADC_ADS124S08_INST_DEFINE(n) ADC_ADS1X4S0X_INST_DEFINE(n, ti_ads124s08, 12, 24)
DT_INST_FOREACH_STATUS_OKAY(ADC_ADS124S08_INST_DEFINE);
#endif
```

This driver is pretty big so it might not be the best one to look at but with that you get the idea I think.

avolmat-st reviewed (changes requested):
Thanks @mjs513.

mjs513 commented:
@avolmat-st  - @KurtE 
Ok finished taking a cut at combining ov7670 and ov7675 libraries.  Also cleaned up the ov7670 library a bit and did some initial testing on the Arduino GIGA display but before that how do you want to take a look at what I got.  I could do a separate PR or just update this PR?

thanks
Mike

mjs513 committed:
Delete ov7670 and ov7675 apis will combine with 7675

mjs513 committed:
Modify cmakelist to use one api for 0v7670 and 0v7675

mjs513 committed:
add combined driver

mjs513 committed:
Resolve SonarCube complaints

avolmat-st commented:
Thanks @mjs513 for the update.

However that is too big update in order to be reviewed or maintained afterward. It is important to have small commits for several reasons:
 - ease of review
 - ability to afterward, in case of issue bisect in the tree in order to figure out which commit introduced an issue and be able to revert it or fix it in another commit.
 - system should be compilable and usable at all commits in order to be able to perform bisect
 - each commit should do one area of modification (as an example, one example of such modification would be to move to the video_cci api on top of the existing driver) and not lots of things at once

Currently there is no need to delete the existing ov7670 driver and introduce a new one while renaming functions, file etc.
It isn't fully clear to me the amount of modifications necessary to make the ov7670 driver work also for ov7675 but they seems to be rather close so my advice it to really take it step by step in order to introduce the ov7675 support within the ov7670.
The filename or functions name is not important at first, when such thing become necessary it is better to have a dedicated commit which does just that (as an example if the filename is simply renamed from ov7670 to ov767x, git will simply say that the file has been renamed).

If this makes the driver smaller and easier to understand, then I'd advice you to start by simply moving the ov7670 driver to the video_cci api. This would make a first commit. Then on top of it make modifications you need in order to introduce the ov7675 compatible within the ov7670 driver.

Regarding the COM10 change, might be better to keep it outside for the time being since it is addressed within the other PR. If you need it in your serie (for testing purpose for example), then you can temporarily have the modification you need in a dedicated commit so that it can be simply removed or updated depending on the result of the discussion in the 2nd PR.

Thank you again for upstreaming all this.

mjs513 committed:
Revert "Resolve SonarCube complaints"

This reverts commit 1e838ef8350ebfec7a3e167964aa4cef5acb9bf6.

mjs513 commented:
@avolmat-st 

In order to do what you want it going to get messy with removing commits.  Going to close this and create new and clean pr with step by step commits

avolmat-st commented:
@mjs513 it isn't necessary to close this PR, you can simply push your updated branch.

mjs513 commented:
> @mjs513 it isn't necessary to close this PR, you can simply push your updated branch.

ok will give it a try but if it gets confusing your fault ---- its a joke

mjs513 committed:
Reverts changes back to just ov7675

josuah commented:
@mjs513
If you wish we can chat about how to do this with git. Feel free to ping me on the Zephyr Discord server
=> https://discord.gg/zephyrproject-720317445772017664

.. or other chat system you typically use

mjs513 committed:
Merge OV7675 into OV7670

deleted left over dups

FIXED TRIPLICATE Driver in ONE FILE

Update OV760 based on GITHUB comments

This set of changes based on comments provide from  avolmat-st  in https://github.com/zephyrproject-rtos/zephyr/pull/94270#pullrequestreview-3114134397.

Tested with OV7675 which works but 7670 framerate is very slow.

mjs513 commented:
> @mjs513 If you wish we can chat about how to do this with git. Feel free to ping me on the Zephyr Discord server => https://discord.gg/zephyrproject-720317445772017664
> 
> .. or other chat system you typically use

Discord it is.  Thanks

avolmat-st commented on drivers/video/ov7670.c:
```diff
@@ -15,15 +19,25 @@
 #include "video_ctrls.h"
 #include "video_device.h"
 
+#if defined(CONFIG_VIDEO_OV7670)
```
Changing the module name might not be really necessary. At the very end if we really want we might change all in ov767x but that's only at the end once everything is in place. For the time being I think it is ok to simply keep only video_ov7670 and consider that this driver is also handling the ov7675

avolmat-st commented on drivers/video/ov7670.c:
```diff
@@ -15,15 +19,25 @@
 	uint8_t cmd;
 };
 
+#if defined(CONFIG_VIDEO_OV7670)
```
Here as well, I don't think there is a need to have separate definition of configs, it seems anyway the 2 structs are exactly the same.
For the same reason I mentioned above regarding the module name, ctrl and data structs can remaining as ov7670_ctrls and ov7670_data.

avolmat-st commented on drivers/video/ov7670.c:
```diff
@@ -33,16 +47,18 @@ struct ov7670_config {
 };
 
+
+#if defined(CONFIG_VIDEO_OV7670)
```
It seems that this patch does not contains everything since the initialization table for the ov7675 are missing. Depending on how different those table are we can see if they need to be split or not.

avolmat-st commented on drivers/video/ov7670.c:
```diff
@@ -189,25 +205,53 @@ const struct ov7670_resolution_cfg OV7670_RESOLUTION_VGA = {
 #define OV7670_MVFP_VFLIP 0x10
 
-#define OV7670_VIDEO_FORMAT_CAP(width, height, format)                                             \
+#define OV767x_VIDEO_FORMAT_CAP(width, height, format)                                             \
```
No need to change the macro name. Also I think this commit doesn't build anymore in OV7670 since the macro OV7670_VIDEO_FORMAT_CAP is still used below

avolmat-st commented on drivers/video/ov7670.c:
```diff
@@ -189,25 +205,53 @@ const struct ov7670_resolution_cfg OV7670_RESOLUTION_VGA = {
+#if defined(CONFIG_VIDEO_OV7670)
 	OV7670_VIDEO_FORMAT_CAP(176, 144, VIDEO_PIX_FMT_RGB565), /* QCIF  */
-	OV7670_VIDEO_FORMAT_CAP(320, 240, VIDEO_PIX_FMT_RGB565), /* QVGA  */
+	OV7670_VIDEO_FORMAT_CAP(176, 144, VIDEO_PIX_FMT_YUYV),   /* QCIF  */
```
Probably unwanted change ? Should add a line instead of changing one if you want to add a format.

avolmat-st commented on drivers/video/ov7670.c:
```diff
@@ -394,7 +440,23 @@ static int ov7670_set_fmt(const struct device *dev, struct video_format *fmt)
+#if defined(CONFIG_VIDEO_OV7675)
+  // Set RGB Format
+  if(fmt->pixelformat == VIDEO_PIX_FMT_RGB565) {
+    ret = video_write_cci_multiregs8(&config->bus, ov7675_rgb565_regs, ARRAY_SIZE(ov7675_rgb565_regs));
```
The whole driver is currently relying on i2c_reg_write_byte_dt calls so we shouldn't mix standard i2c call and video_write_cci calls. Either have a first commit in the serie to move to the video_write_cci framework and then add on top of that the ov7675, either stick to i2c_reg_write_byte_dt call for all.

mjs513 replied:
Thought I would change the ov7670 to video_cci calls in next commit so will do that now since I prefer it that way to be honest.

avolmat-st commented on drivers/video/ov7670.c:
```diff
@@ -458,11 +520,39 @@ static int ov7670_set_fmt(const struct device *dev, struct video_format *fmt)
+			/* Set output format */
+			switch (fmts[i].width_min) {
+			case 160: /* QQVGA */
+        ret = video_write_cci_multiregs8(&config->bus, ov7675_qqvga_regs, ARRAY_SIZE(ov7675_qqvga_regs));
```
Those tables do not exist ..

mjs513 replied:
Yeah thats on me - in my rush forgot to copy them over.  Will fix

avolmat-st commented on drivers/video/ov7670.c:
```diff
@@ -531,6 +625,43 @@ static int ov7670_init(const struct device *dev)
 	}
+#endif
+#else
+
```
Avoid duplicating code. The pwdn / reset control is also in the OV7670 part above,, there are only minor modifications such as polarity.

avolmat-st reviewed (changes requested):
Thanks @mjs513. I did a rapid review of the modifications. My understanding is that only the very last patch has to be reviewed, all previous ones will be dropped. Is that correct ?

josuah reviewed (commented):
Trying to combine the implementation of multiple sensors seems like a good idea.
It will help reducing maintenance efforts for very similar drivers.

But as seen above, this prevents multiple variants to be used at the same time.

It might help to test that the OV7670 and OV7675 support both compile well. Given this is a requirement for merging it, this can save yourself some time by not having to wait the next review.

This is a bit tedious process, but keeping the driver API intact is what permits this to keep working for all the other users (NXP's FRDM-MCXN947, Renesas's new CEU driver), as well as keep the driver well-integrated everywhere (Zephyr Shell support, Webcam support, Video over network support when it comes...).

Thanks for the efforts so far!

avolmat-st commented:
> But as seen above, this prevents multiple variants to be used at the same time.

Not necessarily, it depends on the way it is merged together. I think it is well possible to indicate into the init macro if this is ov7670 or ov7675 and thus hold this into the config / data structure, with pointers as well to the init tables. So, one driver and each instance can run in different mode depending on what was the compatible (aka how it has been initialized).

But first I feel that it is important to understand the difference between the two variants and that's the goal of current activity.

mjs513 commented:
@avolmat-st - @josuah 

Just pushed a comment that I think catches all but the first of your comments.   Going to take me a bit get to the point of fixing module names and driver compatibility issues - need to time to absorb and figure out how it works.

But back to the changes.  I tested with the 7675 and it works no issues.  However, withthe 7670 the framerate is very slow and color is a bit off.  My gut is telling me its an issue with setting registers using his structure vs using tables.  I did try using this 
```
#define OV7670_REG8(addr)	((addr) | VIDEO_REG_ADDR8_DATA8)
```
for the video_write_cci but it seems to have made it worse.

Want to confirm if its alright with you all that I convert the structures for the OV7670 to tables and rewicker the setfmt according to see it that helps - since I know it did when I test the x version

mjs513 commented:
>  me understanding how to do a rebase -i is still a work in progress

mjs513 committed:
Revert "Merge OV7675 into OV7670"

This reverts commit 5abd54f0d70b6fa3ca841761c651b379676fc51f.

mjs513 committed:
Update OV7670 driver to use video cci controls

This commit is focused solely on updating the current OV7670 driver to use the video controls in video_common.   No changes made.

Fix Resutlts of twister run & Compliance check

fix error missing v in video_write_cci register also fixed leading white space but think its wrong

Compliance check fixes & Macro definition

Added in the macro definition for OV7670_REG8 and think fixed the compliance check issues or at least most of them

Added OV7670_REG8 for 8bit regs when using video_cci commands

Based on our discussion I added back in the OV7670_REG8 macro for the OV7670 command registers referrence in video cci commands.

mjs513 commented:
@avolmat-st 

Sorry for the delay in getting this fixed - wanted to make sure camera worked.  Had to add my com10 and reset/pwer pin for testing but that is NOT included in the commit.  Tjos is a copy of git log just that guess we need to start with.
```
commit e29dfebc3f7ab073a90b54798d2d404c088425f5 (HEAD -> Add-OV7675-Driver)
Author: Mike S <CyberMerln@gmail.com>
Date:   Wed Aug 13 15:41:09 2025 -0400

    Update OV7670 driver to use video cci controls

    This commit is focused solely on updating the current OV7670 driver to use the video controls in video_common.   No changes made.
```

> commit 2. if you add new resolution / format to the ov7670 (I've seen this in your file), again, this is another dedicated commit, since this is not related to the ov7675

the additional resolution/format was the result of adding in the ov7675 so do not think this applies

However, after this is done will do another PR to convert over the struct's used for resolutions to tables.  This I think was too much for the commit.

josuah commented:
> In the meantime what should I do with this PR and clean up the branch? revert all the changes back?

One more advantage in working at the [commit level](https://github.com/zephyrproject-rtos/zephyr/pull/94270/commits), rather than the [file level](https://github.com/zephyrproject-rtos/zephyr/pull/94270/files), is that if there are a few commits that are ready, and others that need work, it is possible to merge those that are ready first.

avolmat-st commented on drivers/video/ov7670.c:
```diff
@@ -415,42 +410,45 @@ static int ov7670_set_fmt(const struct device *dev, struct video_format *fmt)
+      ret = video_write_cci_reg(&config->bus, OV7670_COM7,
+               resolution->com7);
+      
+			ret = video_write_cci_reg(&config->bus, OV7670_COM7,
```
video_write_cci_reg / video_read_cci_reg are using specially formatted register values. You can have a look at the header file which describe it in order to let video_write_cci_reg know what kind of register it is. You might want to have a look at other drivers.

avolmat-st commented on drivers/video/ov7670.c:
```diff
@@ -542,24 +539,21 @@ static int ov7670_init(const struct device *dev)
 
-	ret = i2c_write_dt(&config->bus, &cmd, sizeof(cmd));
+
+	ret = video_read_cci_reg(&config->bus, cmd, &pid);
```
Same as for video_read_cci_reg, specially formatted register value is necessary here. uint8_t cmd doesn't seem to be necessary anymore hence should be removed.

avolmat-st reviewed (commented):
Thanks for the update. Few comments added inline.

Can you also run checkpatch.pl prior to push commits, this will highlight comments that will have to be fixed. You can for example run it as follow:

./script/checkpatch.pl -g <the sha1 of the commit>

even better would be
./script/checkpatch.pl -g <the sha1 of the commit> --strict --codespell
However while --strict gives some extract comments, some of them are not always relevant.
--codespell will check spelling and thus help avoid some typo in comments for example (you will have to install a dictionary for that)

mjs513 commented:
Good morning @avolmat-st 

I did run ./scripts/checkpatch.pl drivers/video/ov7670.c and received:
```
drivers/video/ov7670.c:192: WARNING:COMMIT_LOG_LONG_LINE: Possible unwrapped commit description (prefer a maximum 75 chars per line)
#192:
#define OV7670_VIDEO_FORMAT_CAP(width, height, format)                                             \

drivers/video/ov7670.c:633: ERROR:NOT_UNIFIED_DIFF: Does not appear to be a unified-diff format patch

drivers/video/ov7670.c total: 1 errors, 1 warnings, 0 lines checked
```
Understand the first - but line is untouched from original.  Not sure the error message though.  May not be applicable.

With the strict added
```
No codespell typos will be found - file '/usr/share/codespell/dictionary.txt': No such file or directory
3fb0cf48f5ad8aaa5e1df28d2f675c8084b100d2 total: 0 errors, 0 warnings, 0 checks, 99 lines checked

Commit 3fb0cf48f5ad ("drivers: video: fix uninitialized struct") has no obvious style problems and is ready for submission.
```

PS did install dictionary (Codespell) but seems no longer installed to /usr/share/codespell/dictionary.txt but into /usr/lib/python3/dist-packages/codespell_lib/data/dictionary.txt.  Could just copy it but didnt know if you all have a better way.

josuah commented:
Hello @mjs513
I've not forgotten yet though am catching-up with a long list of reviews and other things to fix.
Thanks for the reminder, still.

mjs513 commented:
@avolmat-st, @josuah, @KurtE 
Was finally able to get some of the checkpatch to run using the -f option.  But anyone finding this had to also install the following when using wsl under windows.

```
pip3 install unidiff
pip3 install yamllint
pip3 install junitparser
pip3 install python-magic
pip3 install tabulate
pip3 install lxml
```

KurtE commented:
> Thank you for trying! This should save time while doing PRs as it's possible to check most things locally. Some way to quickly install all Zephyr requirements is to run `west packages pip --install`.

Thanks @josuah @mjs513 @avolmat-st and all

That helped to get it to run better on my windows machine.

mjs513 committed:
drivers: video: ov7670 Update to using i2c commands for PID

drivers: video: ov7670 update to reflect discussion per https://github.com/zephyrproject-rtos/zephyr/discussions/94967

Signed-off-by: Michael Smorto<CyberMerln@gmail.com>

mjs513 commented:
@avolmat-st - @josuah 
I squashed 3 commits into one and added a new commit based on our discussion.  If we all agree I would now like to redo the OV7670_RESOLUTION_QCIF to make use of the video_write_cci_multireg8 command.

josuah commented:
> I squashed 3 commits into one

It might be possible to squash everything in one commit at the end, but do feel free to keep the many commit history as long as this PR is a draft if that helps.

> If we all agree I would now like to redo the OV7670_RESOLUTION_QCIF to make use of the video_write_cci_multireg8 command.

Please, do feel welcome to.

It is great to see the maintenance put on this driver, thanks for the improvements!

mjs513 committed:
drivers: video: ov7670 convert resolution structures to tables.

drivers: video: ov7670 convert resolution structures to tables.

This will make it easier when we merge the ov7675 changes which use tables for its resolution.

Signed-off-by: Michael Smorto<CyberMerln@gmail.com>

mjs513 commented:
@josuah - @avolmat-st 

Quick question - do we still want to merge the ov7675 changes with the ov7670 changes.?

avolmat-st commented on drivers/video/ov7670.c:
```diff
@@ -393,65 +385,33 @@ static int ov7670_set_fmt(const struct device *dev, struct video_format *fmt)
 		}
 		i++;
 	}
+	k_msleep(100);
```
What is the reason for this 100ms delay ?

avolmat-st commented:
> @josuah - @avolmat-st
> 
> Quick question - do we still want to merge the ov7675 changes with the ov7670 changes.?

You mean having the ov7670 driver also supporting ov7675 ? In this case yes.
But in different commit of course.

mjs513 committed:
drivers: video: ov7670 update based on comments

drivers: video: ov7670 update based on comments

This commit incorporates recommended by @avolmat-st of last commit

Signed-off-by: Michael Smorto<CyberMerln@gmail.com>

mjs513 committed:
Update ov7670.c

drivers: video: ov7670 driver changes in prep for OV7675

This commit updates ov7670.c in prep for incorporation of ov7676

Signed-off-by: Michael Smorto<CyberMerln@gmail.com>

mjs513 committed:
drivers: video:  Adds OV7675 driver to OV7670.c

drivers: video: Adds OV7675 driver to OV7670.c

This commit adds OV7675 driver to OV7670.c and updates cmakeslist pointing ov7675 to ov7670.c

Signed-off-by: Michael Smorto<CyberMerln@gmail.com>

mjs513 commented:
@josuah - @avolmat-st  @KurtE(FYI)

The last commit adds in the OV7675 sensor to ov7670.c - not sure if you all want to change the file name or not.

@avolmat-st  -  sorry this last commit can't do much to make it smaller

avolmat-st commented:
Hi @mjs513,

I've seen that you've marked the PR as ready for review. Do you wish to have another run of review on it ?
If that is so, please first prepare a clean PR which can be reviewed by squashing the serie into several commits which introduce modifications unitary enough to be properly reviewed.

Thanks.

mjs513 commented:
@avolmat-st 

Sorry was working on different project and got a bit engrossed in it.  Will close this PR and issue a clean pr.
